llvm/include/llvm/Transforms/Utils/SYCLModuleSplit.h

```diff
@@ -0,0 +1,250 @@
+//===--------- SYCLModuleSplitter.h - split a module into callgraphs
+//-------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Functionality to split a module into call graphs. A callgraph here is a set
+// of entry points with all functions reachable from them via a call. The result
+// of the split is new modules containing corresponding callgraph.
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SYCL_MODULE_SPLIT_H
+#define LLVM_SYCL_MODULE_SPLIT_H
+
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/Error.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+// TODO(maksimsab):
+//  * remove EntryPointGroup::Properties and maybe EntryPointGroup at all.
+//  * perhaps, ModuleDesc::IsTopLevel worth to be removed as well as #ifdef
+//  NDEBUG
+//  * Stub string values for ModuleDesc::Name don't look OK.
+//  * check GenXSPIRVWriterAdaptor comments
+//  * Maybe fix doxygen comments.
+
+namespace llvm {
+
+class Function;
+class Module;
+
+enum IRSplitMode {
+  SPLIT_PER_TU,     // one module per translation unit
```

Comment:
What is our plan for thinLTO? Explicitly tagging @sarnex here. My understanding is that with thinLTO per-source device code split will be performed automatically and from that point of view, per-source mode will be obsolete. Depending on the timelines we may decide to maybe avoid upstreaming of this one at all?

Reply:
sorry, missed this. right, if thinLTO is enabled, everything will be per-TU. However, I'm not convinced thinLTO will always be possible, I think there will be some cases where we can't use it, but I expect those to be few. I don't have a strong opinion on upstreaming per-TU splits, either is fine to me.

Reply:
I am open to supporting only per-kernel mode. I wonder why we wanted to support per-source mode in the first place. If there is no reason, we can just support '-fsycl-device-code-split' (without any argument) and do per-kernel mode all the time.

Thanks